### ISSUE TITLE
False negative: Nullness Checker allows leaking `@UnderInitialization this` to anonymous class instance

### ISSUE THREAD

**Issue by PawelLipski** [typetools/checker-framework#3408]
Pretty similar to #3407, but still a distinct issue (i.e. it's likely that fixing #3407 won't automatically fix this issue and vice versa).

Let's consider the following class:

```java
class SampleClass {
  final String foo;
  String getFoo() { return foo; }

  SampleClass() {
    var anon = new Object() {
      String bar() {
        return SampleClass.this.getFoo().substring(1);
      }
    };
    anon.bar(); /// WHOOPS... NPE, `getFoo()` returns `foo` which is still null
    this.foo = "Hello world";
  }
}
```

This compiles with no error, despite leading to a NPE in the runtime.

**Comment by PawelLipski** [comment]
It's likely that local (non-anonymous) classes are affected as well.

**Comment by mernst** [comment]
Possibly related (but probably not a duplicate): #409
